### ISSUE TITLE
Trailing/regular stop-loss

### ISSUE THREAD

**Issue by csaszkow** [pmorissette/bt#90]
Dear @pmorissette and @sven337!

Thank both of you for your amazing work. I am just getting started with bt and it looks pretty promising. 

After reading through the examples, I am still not sure how to use @sven337's stop-loss function. Here the code I am trying to get the work done:

```
import bt
import pandas as pd

class WeighTarget(bt.Algo):

    def __init__(self, target_weights):
        self.tw = target_weights

    def __call__(self, target):
        # get target weights on date target.now
        if target.now in self.tw.index:
            w = self.tw.ix[target.now]

            # save in temp - this will be used by the weighing algo
            # also dropping any na's just in case they pop up
            target.temp['weights'] = w.dropna()

        # return True because we want to keep on moving down the stack
        return True
    
class StopLoss(bt.Algo):
    """
    Implements a stop loss (regular or trailing). 
    Sells assets if they go below a certain threshold under their maximum value since they were bought.

    Args:
        * threshold (float): threshold of loss that triggers sale. Use 0.05 for 5%.
        * trailing (bool): implement a trailing stop loss if true (use the highest price as reference). 
                           If false, implement a regular stop loss (use the buy price as reference).
        * target_ticker (string): name of the ticker to which to reallocate capital when 
        stoploss orders are triggered. By default, keep proceeds in cash.
    """

    def __init__(self, threshold, trailing=False, target_ticker=None):
        super(StopLoss, self).__init__()
        self.threshold = threshold
        self.trailing = trailing
        self.target_ticker = target_ticker

    def __call__(self, target):
        # For each current position, check its drawdown starting on the date it was purchased

        # If no stoploss was previously set, bail
        if target.perm.get('stoploss_set_up_on') == None:
            target.perm['stoploss_set_up_on'] = {}
            return True

        # Build list of currently held securities
        l = [ target.children[cname] for cname in target.children if target.children[cname]._weight > 0 ]
        for c in l:
            try:
                stoploss_from = target.perm['stoploss_set_up_on'][c.name]
            except KeyError:
                continue
            if not self.trailing:
                # Regular stop loss
                price_base = c.prices.ix[stoploss_from]
                loss = (c.price / price_base) - 1
                triggered = loss < -self.threshold
            else:
                # Trailing stop loss, use ffn's to_drawdown_series
                d = c.prices.ix[stoploss_from:target.now].to_drawdown_series()
                d = d[d < -self.threshold]
                triggered = (len(d) > 0)
                
            if triggered:
                print("Triggering stoploss ",  c.name, target.now, c.price, stoploss_from, c.value)
                reallocate_weight = c.weight
                target.close(c.name)
                if (self.target_ticker):
                    target.rebalance(reallocate_weight, self.target_ticker)

        return True
    
## download some data & calc SMAs
data = bt.get('aapl', start='2008-01-01')

sma_s = data.rolling(30).mean()
sma_l = data.rolling(120).mean()

tw = sma_l.copy()

tw[sma_s > sma_l] = 1.0
tw[sma_s <= sma_l] = -1.0
tw[sma_l.isnull()] = 0.0

tmp = bt.merge(tw, data, sma_s, sma_l)
tmp.columns = ['weights', 'price', 'sma_s', 'sma_l']
ax = tmp.plot(figsize=(15,5), secondary_y=['weights'])

ma_cross = bt.Strategy('ma_cross', [WeighTarget(tw),
                                    bt.algos.Rebalance(),
                                    StopLoss(0.05)])
    


t = bt.Backtest(ma_cross, data)
res = bt.run(t)
print('\n')
res.display()
res.plot()
```

I don't know where else I should put the stop-loss function and also how to customize the Rebalance() class to be able to use the trailing function.

Could you please help out me with this one?

Many thanks in advance!

Cheers,

Mark #

**Comment by sven337** [comment]
StopLoss needs to be run every day, so it has to be at the top of the algo stack. In your proposed code that would probably work as well, however.

As for Rebalance(), you have to modify the end of ```__call__``` that looks like:
```python
        for item in iteritems(targets):
            target.rebalance(item[1], child=item[0], base=base)
```
So that it sets "stoploss_set_up_on" like so:
```python
        for item in iter(targets.items()):
            target.rebalance(item[1], child=item[0], base=base)
            if target.perm.get('stoploss_set_up_on') == None:
                target.perm['stoploss_set_up_on'] = {}
            target.perm['stoploss_set_up_on'][item[0]] = target.now
        return True
```

**Comment by csaszkow** [comment]
Many thanks for your answer, will try it out later today.
